compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/CharHelpers.skiko.kt

```diff
@@ -28,4 +28,77 @@ internal actual fun String.findFollowingBreak(index: Int): Int {
     val it = BreakIterator.makeCharacterInstance()
     it.setText(this)
     return it.following(index)
-}
\ No newline at end of file
+}
+
+internal expect fun isRtlCodePoint(codePoint: Int): Boolean?
+
+/**
+ * Determine direction based on the first strong directional character.
+ * Only considers the characters outside isolate pairs.
+ */
+internal fun String.isRtl(): Boolean? {
+    for (codePoint in codePointsOutsideDirectionalIsolate) {
+        isRtlCodePoint(codePoint)?.let { return it }
+    }
+    return null
+}
+
+/**
+ * U+2066 LEFT-TO-RIGHT ISOLATE (LRI)
+ * U+2067 RIGHT-TO-LEFT ISOLATE (RLI)
+ * U+2068 FIRST STRONG ISOLATE (FSI)
+ */
+private val PUSH_DIRECTIONAL_ISOLATE_RANGE: IntRange = 0x2066..0x2068
+
+/**
+ * U+2069 POP DIRECTIONAL ISOLATE (PDI)
+ */
+private const val POP_DIRECTIONAL_ISOLATE_CODE_POINT: Int = 0x10000
+
+private val String.codePointsOutsideDirectionalIsolate get() = sequence {
+    var openIsolateCount = 0
+    for (codePoint in codePoints) {
+        if (codePoint in PUSH_DIRECTIONAL_ISOLATE_RANGE) {
+            openIsolateCount++
+        } else if (codePoint == POP_DIRECTIONAL_ISOLATE_CODE_POINT) {
+            if (openIsolateCount > 0) {
+                openIsolateCount--
+            }
+        } else if (openIsolateCount == 0) {
+            yield(codePoint)
+        }
+    }
+}
+
+private val String.codePoints get() = sequence {
+    var index = 0
+    while (index < length) {
+        val codePoint = codePointAt(index)
+        yield(codePoint)
+        index += codePoint.charCount()
+    }
+}
+
+private fun String.codePointAt(index: Int): Int {
+    val high = this[index]
+    if (high.isHighSurrogate() && index + 1 < this.length) {
+        val low = this[index + 1]
+        if (low.isLowSurrogate()) {
+            return Char.toCodePoint(high, low)
+        }
+    }
+    return high.code
+}
+
+/**
+ * Converts a surrogate pair to a unicode code point.
+ */
+private fun Char.Companion.toCodePoint(high: Char, low: Char): Int =
+    (((high - MIN_HIGH_SURROGATE) shl 10) or (low - MIN_LOW_SURROGATE)) + 0x10000
```

Comment:
Can you describe this math in doc comment, or add a link to the doc?

Reply:
It's copy-paste from native stdlib (unavailable in common now). See https://github.com/JetBrains/kotlin/blob/3d65420f78e384d2e4f1eeb49b935696a50f0bff/libraries/stdlib/native-wasm/src/kotlin/text/Char.kt#L253-L256

It should be here only until https://youtrack.jetbrains.com/issue/KT-23251

Reply:
Can you please add a TODO comment with these links?

Reply:
Already in https://github.com/JetBrains/compose-multiplatform-core/pull/514/commits/27502f592c313c1c22ae21006246f8d3b4cbdde4